src/components/main/mod.rs

```diff
@@ -24,38 +26,95 @@ pub struct Prop {
 #[allow(non_snake_case)]
 pub fn Main(cx: Scope<Prop>) -> Element {
     log::debug!("rendering Main");
-    let st = use_atom_ref(&cx, STATE).clone();
+    let state = use_atom_ref(&cx, STATE).clone();
     let rg = cx.props.messaging.clone();
-    let state = use_atom_ref(&cx, STATE);
-    let display_welcome = state.read().current_chat.is_none();
-    let sidebar_visibility = match st.read().hide_sidebar {
+    let display_welcome = state.read().selected_chat.is_none();
+    let sidebar_visibility = match state.read().hide_sidebar {
         false => "main-sidebar",
         true => "main-chat",
     };
-    use_future(&cx, (), |_| async move {
-        loop {
-            if let Ok(list) = rg.list_conversations().await {
-                let mut current_conversations: HashMap<Uuid, ConversationInfo> = HashMap::new();
-                for item in &list {
-                    let to_insert = match st.read().all_chats.get(&item.id()) {
-                        Some(v) => v.clone(),
-                        None => ConversationInfo {
-                            conversation: item.clone(),
-                            creation_time: DateTime::from(Local::now()),
-                            ..Default::default()
-                        },
-                    };
-                    current_conversations.insert(item.id(), to_insert);
+
+    use_future(&cx, &rg, |mut rg| async move {
+        log::debug!("streaming conversations");
+
+        // todo: only accept incoming conversations from people we are friends with.
+
+        // get all conversations and update state
+        let mut conversations: HashMap<Uuid, Conversation> = HashMap::new();
+        match rg.list_conversations().await {
+            Ok(r) => {
+                for c in r {
+                    conversations.insert(c.id(), c.clone());
+                }
+            }
+            Err(e) => {
+                log::error!("failed to get conversations: {}", e);
+                return;
+            }
+        }
+
+        // detect removed conversations
+        for id in state.read().all_chats.keys() {
+            if !conversations.contains_key(id) {
+                log::debug!("removing chat");
+                state.write().dispatch(Actions::HideChat(*id));
+            }
+        }
+
+        // detect added conversations
+        for (id, conv) in conversations {
+            if !state.read().all_chats.contains_key(&id) {
+                log::debug!("adding chat");
+                state
+                    .write()
+                    .dispatch(Actions::AddConversation(conv.clone()));
+            }
+        }
+
+        // receive events from Warp
+        let mut stream = loop {
```

Comment:
Would recommend moving this to the top since it also acts as a guard for those 2 errors.

Reply:
good catch. Thank you.